src/java.base/share/classes/java/io/ConsoleImpl.java

```diff
@@ -0,0 +1,359 @@
+     */
+    @Override
+    public char[] readPassword(String fmt, Object ... args) {
+        char[] passwd = null;
+        synchronized (writeLock) {
+            synchronized(readLock) {
+                installShutdownHook();
+                try {
+                    restoreEcho = Console.echo(false);
+                } catch (IOException x) {
+                    throw new IOError(x);
+                }
+                IOError ioe = null;
+                try {
+                    if (!fmt.isEmpty())
+                        pw.format(fmt, args);
+                    passwd = readline(true);
+                } catch (IOException x) {
+                    ioe = new IOError(x);
+                } finally {
+                    try {
+                        if (restoreEcho)
+                            restoreEcho = Console.echo(true);
+                    } catch (IOException x) {
+                        if (ioe == null)
+                            ioe = new IOError(x);
+                        else
+                            ioe.addSuppressed(x);
+                    }
+                    if (ioe != null)
+                        throw ioe;
+                }
+                pw.println();
+            }
+        }
+        return passwd;
+    }
+
+    private void installShutdownHook() {
+        if (shutdownHookInstalled)
+            return;
+        try {
+            // Add a shutdown hook to restore console's echo state should
+            // it be necessary.
+            SharedSecrets.getJavaLangAccess()
+                    .registerShutdownHook(0 /* shutdown hook invocation order */,
+                            false /* only register if shutdown is not in progress */,
+                            new Runnable() {
+                                public void run() {
+                                    try {
+                                        if (restoreEcho) {
+                                            Console.echo(true);
+                                        }
+                                    } catch (IOException x) { }
+                                }
+                            });
+        } catch (IllegalStateException e) {
+            // shutdown is already in progress and readPassword is first used
+            // by a shutdown hook
+        }
+        shutdownHookInstalled = true;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public char[] readPassword() {
+        return readPassword("");
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void flush() {
+        pw.flush();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public Charset charset() {
+        assert CHARSET != null : "charset() should not return null";
+        return CHARSET;
+    }
+
+    private Object readLock;
+    private Object writeLock;
+    private Reader reader;
+    private Writer out;
+    private PrintWriter pw;
+    private Formatter formatter;
```

Comment:
It looks like we can mark `readLock`, `writeLock`, `reader`, `out`, `pw` and `formatter` as `final`.

Reply:
Made them `final` along with other fix-ups.